### PR TITLE
Option to do single precision mesh communication.

### PR DESCRIPTION
**Pull request by atmyers**
This PR adds a number of wrapper functions around the amrex parallel communication routines for mesh data - FillBoundary, ParallelCopy, ParallelAdd, SumBoundary, and OverrideSync, and replaces all uses of these functions with the wrapped versions. It also adds a new runtime option, `warpx.do_single_precision_comms` (default `0`). If this is `0`, the wrapped versions will behave as they currently do on `development`. If this is `1`, the communication will be done using a temporary, single precision version of Multifab instead. Thus, at the cost of some local copies, half as much data will need to be communicated with MPI. This is useful for simulations that use the PSATD solver and thus need a lot of ghost cells.

# To Do

- [x] do not copy data if we already built in single precision
- [x] add docs
- [x] investigate tests that crash from #2319
- [x] add tests
- [x] check if tests fail by a large margin if this would be default-on: #2319

### CODE REVIEW

File: Source/Parallelization/WarpXCommUtil.H
Comment (WeiqunZhang) [inline Source/Parallelization/WarpXCommUtil.H]:
```suggestion
    auto const& srcma = src.const_arrays();
    auto const& dstma = dst.arrays();
    ParallelFor(dst, nghost, numcomp,
    [=] AMREX_GPU_DEVICE (int box_no, int i, int j, int k, int n) noexcept
    {
        dstma[box_no](i,j,k,dstcomp+n) = srcma[box_no](i,j,k,srccomp+n);
    });
    Gpu::synchronize();
```

**Comment by ax3l** [comment]
In #2319, I see some tests change precision with single precision comm (as expected), but I also see quite a lot of tests crashing.

We should investigate the crashes before merging the PR here.

**Comment by ax3l** [comment]
I rebased the experiment in #2319 one last time so we see if no new crashes slurped in

File: Source/Parallelization/WarpXCommUtil.H
Comment (ax3l) [inline Source/Parallelization/WarpXCommUtil.H:7]: Copyright header missing :)

**Review by ax3l — COMMENTED** [review]
Just a minimal comment on a header in a new file.

Otherwise let's wait for CI in this and the linked PR and check it one last time.

**Comment by ax3l** [comment]
You probably saw that we have 3 tests that crash:
https://github.com/ECP-WarpX/WarpX/pull/2319#issuecomment-930350038

They all seem to be related to space charge, which we don't use in the MS.

**Comment by atmyers** [comment]
> You probably saw that we have 3 tests that crash: [#2319 (comment)](https://github.com/ECP-WarpX/WarpX/pull/2319#issuecomment-930350038)
> 
> They all seem to be related to space charge, which we don't use in the MS.

That is expected, I think, due to solver tolerance issues. If we set them looser, those tests pass with `warpx.use_single_precision_comms=1`

File: Source/Parallelization/WarpXCommUtil.H
Comment (atmyers) [reply Source/Parallelization/WarpXCommUtil.H:7]: Added.

**Comment by RemiLehe** [comment]
@atmyers Could you rebase this PR? Thanks!

File: Examples/Modules/boosted_diags/inputs_3d_slice
Comment (ax3l) [inline Examples/Modules/boosted_diags/inputs_3d_slice]: Leftover from local testing, maybe:
```suggestion
btd_openpmd.format = openpmd
```

File: Source/Parallelization/WarpXCommUtil.cpp
Comment (ax3l) [inline Source/Parallelization/WarpXCommUtil.cpp]:
```suggestion
/* Copyright 2021 Andrew Myers
```

File: Source/WarpX.H
Comment (ax3l) [inline Source/WarpX.H:167]:
```suggestion
    //! perform field communications in single precision
    static int do_single_precision_comms;
```

File: Docs/source/usage/parameters.rst
Comment (ax3l) [inline Docs/source/usage/parameters.rst]:
```suggestion
    Perform MPI communications for field guard regions in single precision.
```

**Review by ax3l — APPROVED** [review]

**Pull request merged**
